magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/AspectValidator.scala

```diff
@@ -0,0 +1,126 @@
+      // --- if not set default value is true
+      if(!config.hasPath("validateJsonSchema")) false
+      else config.getBoolean("validateJsonSchema")
+    }
+
+    def validate(aspectId:String, aspectData: JsObject, tenantId: BigInt)(implicit session: DBSession, config: Config) {
+      if(shouldValidate(config)) {
+        AspectPersistence.getById(session, aspectId, tenantId) match {
+          case Some(aspectDef) => validateWithDefinition(aspectDef, aspectData)
+          case None => throw new Exception(s"Failed to validate aspect data: Cannot locate aspect definition for aspect id: ${aspectId}")
+        }
+      }
+    }
+
+    def validateAspects(aspects: Map[String, JsObject], tenantId: BigInt)(implicit session: DBSession, config: Config): Unit = {
+      aspects.foreach( aspect => validate(aspect._1, aspect._2, tenantId))
+    }
+
+    def validateWithDefinition(aspectDef: AspectDefinition, aspectData: JsObject): Unit ={
+      if(!aspectDef.jsonSchema.isDefined) {
+        throw new Exception(s"Failed to validate aspect data: Cannot locate json schema for aspect id: ${aspectDef.id}")
+      }
+      val rawSchema = new JSONObject(aspectDef.jsonSchema.get.toString())
+      val schema = SchemaLoader.load(rawSchema)
+      schema.validate(new JSONObject(aspectData.toString))
+    }
+
+    def validateWithAspectPatch(aspectPatch: JsonPatch, recordId: String, aspectId: String, tenantId: BigInt)(implicit session: DBSession, config: Config): Unit = {
+      val originalAspect = (DefaultRecordPersistence.getRecordAspectById(session, tenantId, recordId, aspectId) match {
+        case Some(aspect) => aspect
+        case None => JsObject()
+      })
+      val patchedAspect = aspectPatch(originalAspect).asJsObject
+      validate(aspectId, patchedAspect, tenantId)(session, config)
+    }
+
+    def validateWithRecordPatch(recordPatch: JsonPatch, recordId: String, tenantId: BigInt)(implicit session: DBSession, config: Config): Unit = {
+      recordPatch.ops
+        .groupBy(
+          op =>
+            op.path match {
+              case "aspects" / (name / _) => Some(name)
+              case _                      => None
+            }
+        )
+        .filterKeys(_.isDefined)
+        .map({
+          case (
+            Some(aspectId),
+            List(Add("aspects" / (_ / rest), aValue))
+            ) =>
+            if (rest == Pointer.Empty) {
+              validate(aspectId, aValue.asJsObject, tenantId)(session, config)
+            } else {
+              validateWithAspectPatch(JsonPatch(Add(rest, aValue)), recordId, aspectId, tenantId)(session, config)
+            }
+          case (
+            Some(aspectId),
+            List(Remove("aspects" / (_ / rest), old))
+            ) =>
+            if(rest != Pointer.Empty) {
+              validateWithAspectPatch(JsonPatch(Remove(rest, old)), recordId, aspectId, tenantId)(session, config)
+            }
+          // We patch in all other scenarios.
```

Comment:
Can we have a comment explaining what scenarios we don't patch in? I'm not sure how the logic works here

Reply:
This logic is copied from here (including the comment 😆):
https://github.com/magda-io/magda/blob/64b58ff8db062cd5a260674514cf9d8cef82da2e/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala#L652

Reason: 
As the validation needs to be done before all patches are processed (therefore, we won't half do the data if one of the patches creates invalid data): 
- we need to apply the full patch to create a patched JSON object from original data. 
- And Then validate on this patched version.
See here: https://github.com/magda-io/magda/blob/64b58ff8db062cd5a260674514cf9d8cef82da2e/magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala#L543

I probably should consolidate this code into a separate file and share with RecordPersistence.scala?

Reply:
Oh right - yeah that'd be great :)

Reply:
I've refactored the code around the patch procedure~